docs/dg/dev/guidelines/performance-guidelines/database-performance-quidelines.md

```diff
@@ -0,0 +1,160 @@
+  - **Update operations** are performed in another separate transaction.
+
+    If an error occurs during the execution of any operation, the corresponding transaction is **rolled back**, and an **exception is thrown**. This ensures data consistency and prevents partial writes in case of failure.
+
+### CascadeActiveRecordBatchProcessorTrait
+
+- Extends `ActiveRecordBatchProcessorTrait` by adding recursive cascade-saving functionality for related entities.
+- Helps manage complex entity relationships during batch processing.
+- Works with `BatchEntityPostSaveInterface`, which enables organizing `postSave` events for related entities in a batch-safe manner.
+
+#### Basic Usage Example
+
+```php
+class AbstractSpySalesOrderItem extends BaseSpySalesOrderItem implements BatchEntityPostSaveInterface
+{
+    // properties re-declared protected to prevent their direct use out from an entity
+    use CascadeActiveRecordBatchProcessorTrait {
+        persist as protected;
+        commit as protected;
+        commitIdentical as protected;
+    }
+    public function batchPostSave(): void
+    {
+        if ($this->statusChanged && $this->getIdSalesOrderItem()) {
+            /** @var \Orm\Zed\Sales\Persistence\SpySalesOrderItem $salesOrderItemEntity */
+            $salesOrderItemEntity = $this;
+            $omsOrderItemStateHistoryEntity = $this->createOmsOrderItemStateHistoryEntity();
+            $omsOrderItemStateHistoryEntity->setOrderItem($salesOrderItemEntity);
+            $omsOrderItemStateHistoryEntity->setFkOmsOrderItemState($this->getFkOmsOrderItemState());
+
+            $this->sharedPersist($omsOrderItemStateHistoryEntity);
+        }
+
+        $this->statusChanged = false;
+    }
+}
+```
+
+With this implementation, the following code:
+
+```php
+class SalesEntityManager extends AbstractEntityManager implements SalesEntityManagerInterface
+{
+    public function updateOrCreateSalesOrderItems(array $salesOrderItemsData): void
+    {
+        // Use batch processor trait for efficient processing
+        foreach ($salesOrderItemsData as $orderItemData) {
+            // Find or create the sales order item
+            $salesOrderItem = $this->findOrCreateSalesOrderItem($orderItemData);
+    
+            // Update item attributes
+            $this->updateSalesOrderItemAttributes($salesOrderItem, $orderItemData);
+    
+            // Collect for batch processing
+            $this->persist($salesOrderItem);
+        }
+    
+        // Commit all collected items in a single batch operation
+    $this->commit();
+    }
+}
+```
+
+Performs the following actions:
+- Loads all required entities from the database in a single query.
+- Creates or updates `salesOrderItem` entities in a single batch.
+- Creates related `OrderItemStateHistoryEntity` instances during the post-save process execution (method `batchPostSave` declared by `BatchEntityPostSaveInterface`) and saves them in a single operation after the main entity save is complete.
```

Comment:
It's not clear why I need to know about the method `batchPostSave`. Rewrite without parenthesis

Reply:
so, what is needed is removing parenthesis?
or you want something else? parenthesis - is explanation, reference of the usage